### DISCUSSION TITLE
get validate Volumes between Storage Group to a link Storage Group

### DISCUSSION THREAD

**Discussion by johndanielk** [dell/PyU4V#202] [General]
I have two storage groups - source and res
I created a link from "SG_source" to..."SG_RES" (the SG_RES is a link from "SG_source"

In the bottom line, I try to make a Validation between an existing Volume name (for example 001CD) and a corresponding Volume name

The question is how can I compare volumes between an existing Storage Group and a Linked Storage Group
See CLI command

```
symsnapvx -sid 111 list -linked
Symmetrix ID              : 11111111111    (Microcode Version: 6079)
--------------------------------------------------------------------------------
Sym                                         Link  Flgs
Dev   Snapshot Name                    Gen  Dev   FCMDS Snapshot Timestamp
----- -------------------------------- ---- ----- ----- ------------------------
001BE sg_PMAX_MYName                     0 001D8 .D.X. Sun Sep 29 14:50:12 2024
001BF sg_PMAX_MYName                     0 001D9 .D.X. Sun Sep 29 14:50:12 2024
001CD sg_PMAX_MYName                     0 001CA .D.X. Sun Sep 29 14:50:12 2024
001D3 sg_PMAX_MYName                     0 001D4 .D.X. Sun Sep 29 14:50:12 2024
```

Here you see that 001CD corresponds to 001CA

Note: I have all the information for both Storage groups of volumes.
I want to know if it is possible through a command like the CLI above to perform the same work only through the rest API

THX

**Comment by rawstorage** [comment]
John,
if you do a get on the snapshot id you used to link you will get the infomration in the lined_storage_group key

https://ipaddress:8443/univmax/restapi/101/replication/symmetrix/000220200226/storagegroup/audit_test/snapshot/HourlyDefault/snapid/76721364737 (GET)

For each volume you will find 
"source_volume_name": "00270",
"linked_volume_name": "00163",


{
  "name": "HourlyDefault",
  "snapid": 76721364737,
  "timestamp": "08:10:05 Mon, 30 Sep 2024 +0000",
  "timestamp_utc": 1727683805000,
  "state": [
    "Established"
  ],
  "num_source_volumes": 2,
  "source_volume": [
    {
      "name": "00270",
      "capacity": 546,
      "capacity_gb": 0.99975586
    },
    {
      "name": "002C5",
      "capacity": 546,
      "capacity_gb": 0.99975586
    }
  ],
  "num_storage_group_volumes": 2,
  "tracks": 0,
  "non_shared_tracks": 0,
  "expired": false,
  "linked": true,
  "restored": false,
  "linked_storage_group_names": [
    "audit_test_LNK_SG_001"
  ],
  "linked_storage_group": [
    {
      "name": "audit_test_LNK_SG_001",
      "source_volume_name": "00270",
      "linked_volume_name": "00163",
      "tracks": 0,
      "trackSize": 131072,
      "percentageCopied": 100,
      "linkedCreationTimestamp": "08:13:48 Mon, 30 Sep 2024 +0000",
      "defined": true,
      "background_define_in_progress": false
    },
    {
      "name": "audit_test_LNK_SG_001",
      "source_volume_name": "002C5",
      "linked_volume_name": "0018A",
      "tracks": 0,
      "trackSize": 131072,
      "percentageCopied": 100,
      "linkedCreationTimestamp": "08:13:48 Mon, 30 Sep 2024 +0000",
      "defined": true,
      "background_define_in_progress": false
    }
  ]
}

**Comment by johndanielk** [comment]
I see you do this through the snapid .
Through snapid it turns out that I have something similar
Let me check please
If necessary, I will write a final script here that will help everyone

**Reply by rawstorage** [reply]
sure, but remember the audience on this GitHub are primarily using PyU4V which has necessary functions already baked, you just need to pull out the keys you need in your logic.

**Comment by johndanielk** [comment]
hi
Unfortunately, I was not able to do it through PyU4V
I'm probably missing the filter thing - I cannot found the filter  like "linked_storage_group" 
this is my simple script in Javascript is some one needs  and all JSON.parse works to build your request

```
var symmetrixId = "000220200226"
var storageGroup = "audit_test";
var snapshotName = "audit_test_snapName";
var snapId = "76721364737";

var url = "/univmax/restapi/101/replication/symmetrix/" + symmetrixId +
          "/storagegroup/" + storageGroup + "/snapshot/" + snapshotName + "/snapid/" + snapId;
          
var req = pmaxHost.createRequest("GET", url, null);
var authHeader = "Basic " + basicAuto;
req.setHeader("Authorization", authHeader);
req.setHeader("Accept", "application/json");
req.setHeader("Content-Type", "application/json");

var res = req.execute();
if (res.statusCode != 200) {
    System.log("~~~~ Error: Couldn't retrieve snapshot details ~~~");
    throw("Couldn't retrieve snapshot details: " + res.contentAsString);
}

var snapshotDetails = JSON.parse(res.contentAsString);
System.log("Snapshot Details: " + JSON.stringify(snapshotDetails, null, 2));
```
note, 
if you can share a screen shots how can you build like PyU4V will be great

THX

**Comment by rawstorage** [comment]
john,
I actually had a blog on exactly this some time ago. link below

https://rawstorage.wordpress.com/2023/10/17/resting-on-our-laurels-whats-new-with-the-powermax-api-and-unisphere-10-1/#:~:text=The%20ability%20to%20list%20storage%20groups%20that%20are%20linked%20targets%20has%20been%20there%20for%20a%20while%2C%20with%20query%20parameters%20in%20the%20following%20GET%20call%3A

Corresponding functions in PyU4V are 

PyU4V.replication.get_replication_enabled_storage_groups(is_link_target=true)
PyU4V.replication.get_storage_group_replication_details()
PyU4V,replication.get_snapshot_snap_id_details()

Please check the documentation on https://pyu4v.readthedocs.io/en/latest/PyU4V.html#module-PyU4V.replication if you are stuck.

If I get some free time I might be able to knock up a working example but you should have the information here to do everything you need,

**Comment by rawstorage** [comment]
Quick and dirty in PyU4V

import PyU4V

conn = PyU4V.U4VConn(username='smc', password='smc',
                    server_ip='ipaddress', verify=None,
                    array_id='000220200226', timeout=300)

print(conn.common.get_uni_version())

linked_sg_list = conn.replication.get_replication_enabled_storage_groups(
    is_link_target=True)

for sg in linked_sg_list:
    snap_details = conn.replication.get_storage_group_replication_details(sg)
    src_sg = snap_details['snapshot_link_info']['source_storage_group_names']
    snap_name = snap_details['snapshot_link_info']['snapshot_name']
    snap_id = snap_details['snapshot_link_info']['snap_id']
    link_details = conn.replication.get_snapshot_snap_id_details(
        sg_id=src_sg[0], snap_name=snap_name, snap_id=snap_id)[
        'linked_storage_group']

    print(f'Source sg = {src_sg}', f'Snapshot name = {snap_name}',
          f'Snapshot ID = {snap_id}', f'Linked SG = {sg}')

    volume_pairings = [f'Source Dev {link["source_volume_name"]} Paired with {link["linked_volume_name"]}' for link in link_details]
    print('volume pairings')
    print('\n'.join(volume_pairings))

Output running

![image](https://github.com/user-attachments/assets/a1eda827-d511-4d06-abd3-6a262dbabbfc)

**Reply by rawstorage** [reply]
not sure what's going on in the formatting above.. But it works
